pkg/webhook/deploy/cert/secret.go

```diff
@@ -31,11 +23,62 @@ const (
 	// CACert is the name of the key associated with the certificate of the CA for
 	// the keypair.
 	CACert = "ca-cert.pem"
+
+	// certSecretName is a name of the secret
+	certSecretName = "webhook-certs"
+
+	// Time used for updating a certificate before it expires.
+	oneWeek = 7 * 24 * time.Hour
+
+	// serviceName is the name of webhook service
+	serviceName = "member-operator-webhook"
 )
 
-// CreateSecret creates a secret containing certificate data
-func CreateSecret(name, namespace, serviceName string) (*corev1.Secret, error) {
-	serverKey, serverCert, caCert, err := CreateCerts(serviceName, namespace, time.Now().AddDate(1, 0, 0))
+func EnsureSecret(cl client.Client, namespace string, expiration time.Time) ([]byte, error) {
+	certSecret := &corev1.Secret{}
+	if err := cl.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: certSecretName}, certSecret); err != nil && !errors.IsNotFound(err) {
+		return nil, err
+	} else if err != nil {
+		// does not exist, so let's create it
+		certSecret, err := newSecret(certSecretName, namespace, serviceName, expiration)
+		if err != nil {
+			return nil, err
+		}
+		if err := cl.Create(context.TODO(), certSecret); err != nil {
+			return nil, err
+		}
+		return certSecret.Data[CACert], nil
+	}
+
+	// already exists - check the expiration date of the certificate to see if it needs to be updated
+	cert, err := tls.X509KeyPair(certSecret.Data[ServerCert], certSecret.Data[ServerKey])
+	if err != nil {
+		log.Error(err, "creating pem from certificate and key failed")
+	} else {
+		certData, err := x509.ParseCertificate(cert.Certificate[0])
+		if err != nil {
+			log.Error(err, "parsing certificate failed")
+		} else if time.Now().Add(oneWeek).Before(certData.NotAfter) {
```

Comment:
We check/create the certs only on the webhook/operator startup, correct? It means that if there is no updates/re-starts of the operator within the last week of expiration timeout then we will get invalid certs used.

If we still want to use that timeout let's use some timeout which is a half of the expiration timeout.
For example it the TTL is 12 months than the timeout is 6 month. It would be also nice to have both of these const declared near each other so we see the relation. Or we could just dynamically calculate the timeout based on the expiration const.

Later we could introduce some controller which would re-queue every N days to ensure the certs are up to date.

Reply:
> We check/create the certs only on the webhook/operator startup, correct? It means that if there is no updates/re-starts of the operator within the last week of expiration timeout then we will get invalid certs used.

yes, you are right. I realized this as well, but I thought that later I would most likely add a controller (as you also mentioned in the comment) that would watch all objects of the template - in the same way as we do with reg-service. With the controller and postponed reconcile loops we should be able to ensure that there is at least one re-queue within a day (or a few days, as you mentioned as well)

>If we still want to use that timeout let's use some timeout which is a half of the expiration timeout.
For example it the TTL is 12 months than the timeout is 6 month. 

That's very good point, having the timeout way longer will easily solve this potential issue. No idea why I didn't realize it :man_shrugging:  let me do it :wink:

Reply:
done here 54fd990
the secrets are replaced after half of the expiration timeout